java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/launch/NbProcessConsole.java

```diff
@@ -71,7 +71,18 @@ public void stop() {
     protected InputStream getStdIn() throws IOException {
         synchronized (this) {
             if (inputSink == null) {
-                inputSink = new PipedInputStream(inputSource);
+                inputSink = new PipedInputStream(inputSource) {
+                    @Override
+                    public void close() throws IOException {
+                        synchronized(this) {
```

Comment:
Why is this explicit synchronization necessary? There is synchronize already in super.close() and in `receivedLast()` which is called from `inputSource.close()`.

Reply:
This way, the internal `closedByReader` flag + `closedByWriter` flag will be set 'atomically' for the blocking reader and only then it is allowed to run (acquires its lock). Not sure if it matters, in fact, but I wanted to have the close 'consistent'.

Reply:
O.K.